pkg/controllers/capr/managesystemagent/managesystemagent.go

```diff
@@ -465,34 +512,118 @@ func CurrentVersionResolvesGH5551(version string) bool {
 	return curSemVer.GreaterThanEqual(GH5551FixedVersions[int(minor)])
 }
 
-func ToResources(objs []runtime.Object) (result []v1alpha1.BundleResource, err error) {
-	for _, obj := range objs {
-		obj = obj.DeepCopyObject()
-		if err := gvk.Set(obj); err != nil {
-			return nil, fmt.Errorf("failed to set gvk: %w", err)
-		}
+// UninstallFleetBasedApps handles the removal of Fleet Bundles for both the managed-system-agent and
+// managed-system-upgrade-controller, and also takes care of removing the AppliedSystemAgentUpgraderHashAnnotation
+// annotation in a specific scenario.
+func (h *handler) UninstallFleetBasedApps(_ string, cluster *rancherv1.Cluster) (*rancherv1.Cluster, error) {
+	if cluster == nil || cluster.DeletionTimestamp != nil {
+		return cluster, nil
+	}
+	if cluster.Spec.RKEConfig == nil {
+		return cluster, nil
+	}
+	// The absence of the FleetWorkspaceName indicates that Fleet is not ready on the cluster, so do Fleet bundles
+	if cluster.Status.FleetWorkspaceName == "" {
+		return cluster, nil
+	}
+
+	// Skip if Rancher does not have a connection to the cluster
+	if !clusterconnected.Connected.IsTrue(cluster) {
+		return cluster, nil
+	}
+
+	// Skip if the cluster's kubeconfig is not populated
+	if cluster.Status.ClientSecretName == "" {
+		return cluster, nil
+	}
 
-		typeMeta, err := meta.TypeAccessor(obj)
+	if settings.SystemAgentUpgradeImage.Get() == "" {
+		logrus.Debugf("[managesystemagent] cluster %s/%s: the SystemAgentUpgradeImage setting is not set, skip uninstalling Fleet-based apps", cluster.Namespace, cluster.Name)
+		return cluster, fmt.Errorf("[managesystemagent] cluster %s/%s: the SystemAgentUpgradeImage setting is not set", cluster.Namespace, cluster.Name)
+	}
+	if settings.SystemUpgradeControllerChartVersion.Get() == "" {
+		logrus.Debugf("[managesystemagent] cluster %s/%s: the SystemUpgradeControllerChartVersion setting is not set, skip uninstalling Fleet-based apps", cluster.Namespace, cluster.Name)
+		return cluster, fmt.Errorf("[managesystemagent] cluster %s/%s: the SystemUpgradeControllerChartVersion setting is not set", cluster.Namespace, cluster.Name)
+	}
+
+	dropAnnotation := false
+	defer func() {
```

Comment:
I wouldn't put this here. I would just move this down to line 616, and not use a defer.

Reply:
fixed